### PR TITLE
feat: add automatic conflict resolution support for delete operations

### PR DESCRIPTION
**Pull request by wjones127**
## Summary

Add `affected_rows` tracking to delete operations to enable automatic conflict resolution similar to update operations. This allows concurrent delete operations to succeed with proper retry-based conflict resolution.

## Changes

- **Add affected_rows field to DeleteData struct** - Stores the `RowIdTreeMap` containing row addresses affected by the delete operation
- **Capture RowIdTreeMap during delete execution** - Convert `removed_row_addrs` to `RowIdTreeMap::from()` for conflict detection
- **Pass affected_rows to CommitBuilder.with_affected_rows()** - Enable automatic conflict resolution in commit step
- **Add test_delete_concurrency test** - Similar to `test_update_concurrency`, verifies concurrent delete operations work correctly
- **Handle edge cases** - Proper handling for delete-all and delete-nothing scenarios

## Test Plan

- [x] All existing delete tests pass
- [x] New `test_delete_concurrency` test passes with 3 concurrent delete operations
- [x] `test_concurrent_delete_with_retries` continues to work with overlapping operations
- [x] Verified compatibility with retry-based conflict resolution infrastructure

Closes https://github.com/lancedb/lance/issues/3796

🤖 Generated with [Claude Code](https://claude.ai/code)

### CODE REVIEW

File: rust/lance/src/dataset/write/delete.rs [hunk 1]
```diff
@@ -183,7 +185,9 @@ impl RetryExecutor for DeleteJob {
                         .iter()
                         .map(|f| f.id() as u64)
                         .collect();
-                    (Vec::new(), deleted_fragment_ids)
+                    // When deleting everything, we don't have specific row addresses,
+                    // but we can create an empty RowIdTreeMap since all fragments are deleted
+                    (Vec::new(), deleted_fragment_ids, RowIdTreeMap::new())
```
Comment (jackye1995) [inline rust/lance/src/dataset/write/delete.rs]: I think the logic should be right here based on the `conflict_resolver` logic, but can we add a test to confirm?

**Review by jackye1995 — APPROVED** [review]
looks good to me, just a comment regarding testing

File: rust/lance/src/dataset/write/delete.rs (hunk 1 above)
Comment (wjones127) [reply rust/lance/src/dataset/write/delete.rs]: Good catch. Added a test and modified this not to return an `affected_rows` here.
